### PR TITLE
can_solve

### PR DESCRIPTION
**Pull request by hwjsnc**
This Pull request is a step towards solving issue #635 

The new functions can_solve and can_solve_with_solution are not efficient for `side = :right`. If the API is ok so far, I will continue working on the implementation.

### CODE REVIEW

**Comment by thofma** [comment]
I am happy with the API.

**Comment by wbhart** [comment]
Yes, the interface here looks good.

**Comment by wbhart** [comment]
I see a lot has been done since the initial interface. Is this in a state you are happy with hwjsnc? We are looking at merging it as soon as possible when you are happy for us to do so.

I will take a closer look later today, but I notice so far you have markdown docstrings, but these are not pulled into the docs anywhere in docs/src/matrix.md unless I am overlooking that. Is it possible you could do this and add some examples there?

**Comment by wbhart** [comment]
@rfourquet There hasn't been a reply from @hwjsnc so I think we can perhaps take over this PR and get it merged?

**Comment by rfourquet** [comment]
>  I think we can perhaps take over this PR and get it merged?

Agreed. There are a couple small things to fix here, but I can issue a PR right after merging this one.

Also, I know that `AbstractAlgebra` already have `can_solve_left_reduced_triu`, and that `Hecke` already implemented `can_solve` methods, but was wondering how settled this name is. @hwjsnc suggested the possibility to use a union-type in https://github.com/Nemocas/AbstractAlgebra.jl/issues/635#issue-687441467, and I slightly lean towards this solution, with a `trysolve` function (and `tryinvert`), over `can_solve` (and `invertible_with_inverse`), i.e. return the result or `nothing` if not possible. One small advantage is that if there is no result, you don't have to allocate a dummy value like in `can_solve` or `invertible_with_inverse`.

**Comment by hwjsnc** [comment]
@wbhart @rfourquet Sorry about the silence, I was busy with other things. I am generally happy with the state of this PR.

I left the docs unchanged because I had assumed they would be created automatically from the docstrings. I can fix them later today.

A note on performance: The code still computes `side=:right` in terms of `side=:left` (with some transposes). I had considered handling this case as well, a simple [benchmark](https://gist.github.com/hwjsnc/75a2021c4dfce2ae8e737fde6d9d3148) suggests that the transposes have no significant negative effect on the runtime. However, the results also suggest that for general rings, the performance _with_ transposes is better than without for moderately large (≥ 20×20) matrices, so I'm not sure the benchmark is reliable.

**Comment by thofma** [comment]
Please no union types. I don't think there is a need to allocate a dummy variable, since you can just return the input matrix.

**Comment by rfourquet** [comment]
> Please no union types. I don't think there is a need to allocate a dummy variable, since you can just return the input matrix.

Indeed the returning the input matrix would do here, @hwjsnc could you do that in this PR?

@thofma why not union types? You want to avoid non-type-stability, or you don't like the corresponding API?

File: src/generic/Matrix.jl
Comment (rfourquet) [inline src/generic/Matrix.jl:2336]:
```suggestion
                                                                                side::Symbol = :right) where S <: RingElement
```
Also, as you have a newline here, you could align `side` more on the left, with `a` for example.

File: src/generic/Matrix.jl
Comment (rfourquet) [inline src/generic/Matrix.jl:2328]:
```suggestion
> Given two matrices $a$ and $b$ over the same ring, try to solve $ax = b$
```

File: src/generic/Matrix.jl
Comment (rfourquet) [inline src/generic/Matrix.jl:2330]:
```suggestion
> return a tuple `(flag, x)`. If a solution exists, `flag` is set to true and
```

File: src/generic/Matrix.jl
Comment (rfourquet) [inline src/generic/Matrix.jl:2373]:
```suggestion
      error("Unsupported argument :$side for side: must be :left or :right.")
```

File: src/generic/Matrix.jl
Comment (rfourquet) [inline src/generic/Matrix.jl:2378]:
```suggestion
                                                                                side::Symbol = :right) where T <: FieldElement
```

File: src/generic/Matrix.jl
Comment (rfourquet) [inline src/generic/Matrix.jl:2413]:
```suggestion
> Given two matrices $a$ and $b$ over the same ring, check the solubility
```

File: src/generic/Matrix.jl
Comment (rfourquet) [inline src/generic/Matrix.jl:2415]:
```suggestion
> Return true if a solution exists, false otherwise. If the dimensions
```

File: src/generic/Matrix.jl
Comment (rfourquet) [inline src/generic/Matrix.jl:2472]:
```suggestion
> Given an $n\times m$ matrix $A$ over a ring, return a tuple `(flag, B)`.
```

File: src/generic/Matrix.jl
Comment (rfourquet) [inline src/generic/Matrix.jl:2476]:
```suggestion
> If `flag` is false, no right or left inverse does exist.
```

File: src/generic/Matrix.jl
Comment (rfourquet) [inline src/generic/Matrix.jl:2492]:
```suggestion
> Return true if a given square matrix is invertible, false otherwise. If
```

File: test/generic/Matrix-test.jl
Comment (rfourquet) [inline test/generic/Matrix-test.jl:1612]: Maybe you can test also `can_solve_with_solution` with `side = :right` ? And test that an error is thrown when `side` is not `:right` or `:left`.

**Comment by thofma** [comment]
It is a mix of both. I dislike cluttering the code base with `if n === nothing`. It also slows down compilation/inference.

**Comment by rfourquet** [comment]
>  I dislike cluttering the code base with if n === nothing

Ok; instead there will still be `if flag`, but it's indeed slightly less verbose.

> It also slows down compilation/inference.

I hadn't thought about that (and never tried to measure it).

File: src/generic/Matrix.jl
Comment (hwjsnc) [reply src/generic/Matrix.jl:2336]: I wanted to emulate the formatting in [Hecke.jl](https://github.com/thofma/Hecke.jl/blob/master/src/Misc/Matrix.jl#L1822-L1823), but now I see that in AbstractAlgebra.jl, the function signature is generally all in the same line. I personally prefer it this way too, so if it's ok I would change it like this:

```
- function can_solve_with_solution(a::AbstractAlgebra.MatElem{S}, b::AbstractAlgebra.MatElem{S};
-                                                                                 side = :right) where S <: RingElement
+ function can_solve_with_solution(a::AbstractAlgebra.MatElem{S}, b::AbstractAlgebra.MatElem{S}; side::Symbol = :right) where S <: RingElement 
```

(Sorry about the formatting, I can't figure out how to do it with "suggestion")

File: src/generic/Matrix.jl
Comment (hwjsnc) [reply src/generic/Matrix.jl:2476]: I'd prefer: ```If `flag` is false, no right or left inverse exists.```

File: test/generic/Matrix-test.jl
Comment (hwjsnc) [reply test/generic/Matrix-test.jl:1612]: What do you suggest to test for `side`? Something nonsensical like `side=:aaa`, or a random symbol, or something else?

Comment (rfourquet) [reply test/generic/Matrix-test.jl:1612]: Yes either `:aaa` or `Symbol(randstring())` will do. And why not with something else, like the string `"right"`.

File: src/generic/Matrix.jl
Comment (rfourquet) [reply src/generic/Matrix.jl:2336]: Sure do it as you prefer. I don't think there is convention in AbstractAlgebra regarding line lenghts, I like it not too long though (.e.g less than 92 characters) as it makes it more difficult to read when one's screen is not that wide.

**Comment by hwjsnc** [comment]
@rfourquet 

> > Please no union types. I don't think there is a need to allocate a dummy variable, since you can just return the input matrix.
> 
> Indeed the returning the input matrix would do here, @hwjsnc could you do that in this PR?

Where do you mean? As far as I can tell the variables are all necessary, though I could have missed something of course.

File: src/generic/Matrix.jl
Comment (rfourquet) [inline src/generic/Matrix.jl:2369]: > Where do you mean? As far as I can tell the variables are all necessary,

@hwjsnc I meant re-using the input matrix, like `a` here for example, instead of `zero(a, 0, 0)`, to avoid allocating a dummy matrix.

**Comment by wbhart** [comment]
Thanks @hwjsnc 

I think this can be merged when @rfourquet is happy. However, there might be consequences for Hecke. Should we mark this as breaking @thofma ?

**Comment by hwjsnc** [comment]
Suggestions from @rfourquet's code review are now incorporated and the docs are updated.

I haven't eliminated the matrix allocations in the error cases yet. I think that returning the input matrix would make it easier to misuse the function. Imagine a user calling `can_solve_with_solution` for two n×n matrices and forgetting to check the flag when there is no solution, but using the returned matrix without apparent error. When a 0×0 matrix is returned that wouldn't be the case. The performance impact should also be minimal compared to computation of hnf/rr.

**Comment by thofma** [comment]
This looks great, thanks and I am happy with it.

If I understand correctly, there was no `can_solve` exported from  AbstractAlgebra before. So this should not be breaking for Hecke.

**Review by rfourquet — APPROVED** [review]
Thanks @hwjsnc ! Looks like in a good state to merge now. And the thorough testing is very nice too, together with the added examples.

**Comment by rfourquet** [comment]
The tests fail only on Julia nightlies, with a know error (seen elsewhere).

**Comment by hwjsnc** [comment]
Thanks for the merge everyone!

**Pull request merged**
